e2e_tests/docker_image/Dockerfile

```diff
@@ -0,0 +1,82 @@
+FROM ubuntu:18.04
+
+ENV PKG_CONFIG_PATH /usr/local/lib/pkgconfig
+
+RUN apt update
+RUN apt install -y autoconf-archive libcmocka0 libcmocka-dev procps
+RUN apt install -y iproute2 build-essential git pkg-config gcc libtool automake libssl-dev uthash-dev doxygen libjson-c-dev
+RUN apt install -y --fix-missing wget python3 cmake clang
+RUN apt install -y gcc-multilib
+RUN apt install -y gcc-arm-linux-gnueabihf
+RUN apt install -y gcc-aarch64-linux-gnu
+RUN apt install -y libini-config-dev libcurl4-openssl-dev tpm2-tools curl libgcc1
+
+WORKDIR /tmp
+
+# Download and install TSS 2.0
+RUN git clone https://github.com/tpm2-software/tpm2-tss.git --branch 2.3.3
+RUN cd tpm2-tss \
+	&& ./bootstrap \
+	&& ./configure \
+	&& make -j$(nproc) \
+	&& make install \
+	&& ldconfig
+
+RUN rm -rf tpm2-tss
+
+ARG ibmtpm_name=ibmtpm1637
+
+# Download and install software TPM
+RUN wget -L "https://downloads.sourceforge.net/project/ibmswtpm2/$ibmtpm_name.tar.gz"
+RUN sha256sum $ibmtpm_name.tar.gz | grep ^dd3a4c3f7724243bc9ebcd5c39bbf87b82c696d1c1241cb8e5883534f6e2e327
+RUN mkdir -p $ibmtpm_name \
+	&& tar xvf $ibmtpm_name.tar.gz -C $ibmtpm_name \
+	&& rm $ibmtpm_name.tar.gz
+WORKDIR $ibmtpm_name/src
+RUN sed -i 's/-DTPM_NUVOTON/-DTPM_NUVOTON $(CFLAGS)/' makefile
+RUN CFLAGS="-DNV_MEMORY_SIZE=32768 -DMIN_EVICT_OBJECTS=7" make -j$(nproc) \
+&& cp tpm_server /usr/local/bin
+RUN rm -rf $ibmtpm_name/src $ibmtpm_name.tar.gz
+
+
+# Download and install SoftHSMv2
+WORKDIR /tmp
+RUN git clone https://github.com/opendnssec/SoftHSMv2.git \
+	&& cd SoftHSMv2 \
+	&& git reset --hard 20a53bd083a6134ce2230f80edda5dc8be0366bd
+
+RUN cd SoftHSMv2 \
+	&& sh autogen.sh \
+	# Autogen might fail because of some limitations of autoconf, see:
+	# https://bugzilla.redhat.com/show_bug.cgi?id=1826935#c3
+	|| sh autogen.sh
+RUN cd SoftHSMv2 \
+	&& ./configure --disable-gost \
+	&& make \
+	&& make install
+RUN rm -rf SoftHSMv2
+
+# Create a new token in a new slot. The slot number assigned will be random
+# and is found with the find_slot_number script.
+RUN softhsm2-util --init-token --slot 0 --label "Parsec Tests" --pin 123456 --so-pin 123456
+
+# Install Rust toolchain for all users
+# This way of installing allows all users to call the same binaries, but non-root
+# users cannot modify the toolchains or install new ones.
+# See: https://github.com/rust-lang/rustup/issues/1085
+ENV RUSTUP_HOME /opt/rust
+ENV CARGO_HOME /opt/rust
+RUN curl https://sh.rustup.rs -sSf | sh -s -- -y --no-modify-path
+ENV PATH="/root/.cargo/bin:${PATH}"
+
+# Install the wrappers for the Rust binaries installed earlier
+COPY _exec_wrapper /usr/local/bin/
+RUN ls /opt/rust/bin | xargs -n1 -I% ln -s /usr/local/bin/_exec_wrapper /usr/local/bin/$(basename %)
+
+# Adding the targets for the cross-compilation tests
+RUN rustup target add armv7-unknown-linux-gnueabihf
+RUN rustup target add aarch64-unknown-linux-gnu
+
+# Add users for multitenancy tests
+RUN useradd -m parsec-client-1
+RUN useradd -m parsec-client-2
```

Comment:
Maybe a:
```
RUN apt autoremove
```
or something similar would slim the container even more?

Reply:
Just tried running it inside the built container - only removed about 60 MB of stuff. Will add it, though. I might try and look up which of the dependencies are no longer needed, maybe we can slim some of that too, though I think the big ones are needed (gcc, clang...). Do we need Python for any "runtime compilation" step? 🤔

Reply:
ah yes true, maybe some dependencies are only needed to build things but then not needed anymore? Python is heavily used by Ubuntu, last time I uninstalled it on my system, it destroyed everything 🔥

Reply:
Hmm, yeah, I've been prodding here and there, but don't think there's much space I can save.. :( The libraries that I can remove are small, around 1MB.

Reply:
Initially I had added this, but then things started failing because some useful libraries were being removed... So I've scrapped it again, but it's good to know that exists